### PR TITLE
guard ourshelves from being called after package close at VS shutdown

### PR DESCRIPTION
**Pull request by heejaechang**
contract is vs package should not call other component on vs shutdown. but we should guard ourselves from other package violating that contract.

### CODE REVIEW

**Comment by heejaechang** [comment]
@tmat @pharring @jasonmalinowski can you take a look?

**Comment by heejaechang** [comment]
@KevinH-MS @tmat @wschae can one of you tell debugger team not to call us at VS shutdown? (their package close)

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs [hunk 1]
```diff
@@ -20,57 +19,58 @@ public int HasCustomMetadataEmitter(out bool value)
 
         public int StartDebuggingPE()
         {
-            return EditAndContinueImplOpt.StartDebuggingPE();
+            return EditAndContinueImplOpt != null ? EditAndContinueImplOpt.StartDebuggingPE() : VSConstants.S_OK;
```
Comment (pharring) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs]: Can be written as

    return EditAndContinue?.StartDebuggingPE() ?? VSConstants.S_OK;

**Comment by pharring** [comment]
:+1:

**Comment by KevinH-MS** [comment]
:+1:

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs [hunk 2]
```diff
@@ -20,57 +19,58 @@ public int HasCustomMetadataEmitter(out bool value)
...
         }
 
         public int StopDebuggingPE()
         {
-            return EditAndContinueImplOpt.StopDebuggingPE();
+            return EditAndContinueImplOpt != null ? EditAndContinueImplOpt.StopDebuggingPE() : VSConstants.S_OK;
```
Comment (tmat) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs]: Shouldn't all of these rather return E_FAIL?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs [hunk 3]
```diff
@@ -20,57 +19,58 @@ public int HasCustomMetadataEmitter(out bool value)
...
         }
 
         public int GetPEidentity(Guid[] pMVID, string[] pbstrPEName)
         {
-            return EditAndContinueImplOpt.GetPEidentity(pMVID, pbstrPEName);
+            return EditAndContinueImplOpt != null ? EditAndContinueImplOpt.GetPEidentity(pMVID, pbstrPEName) : VSConstants.S_OK;
```
Comment (tmat) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs]: Especially this one - pMVID and pbstrPEName are expected to be filled with values.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs [hunk 4]
```diff
@@ -20,57 +19,58 @@ public int HasCustomMetadataEmitter(out bool value)
...
         public int GetExceptionSpans(uint celt, ShellInterop.ENC_EXCEPTION_SPAN[] rgelt, ref uint pceltFetched)
         {
-            return EditAndContinueImplOpt.GetExceptionSpans(celt, rgelt, ref pceltFetched);
+            return EditAndContinueImplOpt != null ? EditAndContinueImplOpt.GetExceptionSpans(celt, rgelt, ref pceltFetched) : VSConstants.S_OK;
         }
 
         public int GetCurrentExceptionSpanPosition(uint id, VsTextSpan[] ptsNewPosition)
         {
-            return EditAndContinueImplOpt.GetCurrentExceptionSpanPosition(id, ptsNewPosition);
+            return EditAndContinueImplOpt != null ? EditAndContinueImplOpt.GetCurrentExceptionSpanPosition(id, ptsNewPosition) : VSConstants.S_OK;
```
Comment (tmat) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs]: The same here.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs [hunk 5]
```diff
@@ -20,57 +19,58 @@ public int HasCustomMetadataEmitter(out bool value)
...
         }
 
         public int GetENCBuildState(ShellInterop.ENC_BUILD_STATE[] pENCBuildState)
         {
-            return EditAndContinueImplOpt.GetENCBuildState(pENCBuildState);
+            return EditAndContinueImplOpt != null ? EditAndContinueImplOpt.GetENCBuildState(pENCBuildState) : VSConstants.S_OK;
```
Comment (tmat) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs]: And here.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs (hunk 1 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs]: done

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs (hunk 2 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs]: caller might care it being failed but probably won't care if it succeeded even if we didn't actually do anything at shutdown.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs (hunk 3 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractEncProject_IVsENCRebuildableProjectCfg.cs]: I changed everything that has out param to return E_Fail.

**Comment by heejaechang** [comment]
@srivatsn this is RC bug fix. I will prepare RC PR.

**Pull request merged**